design/StableMemory.md

```diff
@@ -144,90 +153,115 @@ If there are no stable variables, we shift the first two words of
 StableMemory to the end, and simply write the 32-bit word 0x0,
 followed by the 32-bit size of `max`, of StableMemory.
 
-In post_upgrade, we reverse this process do discover any stable vars and size of StableMemory,
+In post_upgrade, we reverse this process to discover any stable vars and size of StableMemory,
 taking care to zero the memory vacated by shifting back the initial portion of StableMem
 (so that reads beyond `max` see what they would expect).
 
 Stable memory layout (during execution):
 
-*  aligned with stable-memory, with global max holding max size of memory read or written in i32 (initially 0).
-*  each write adjusts max_size and grows memory if necessary
-*  each read adjusts max_size and grows memory if necessary, reading 0 for fresh locations.
+*  aligned with stable-memory, with global word `size` holding logical page count (initially 0 < !size < 2^16).
+*  user are responsible for allocating logical pages.
+*  each load/store does a `size`-related bounds check.
+
+(stable variables aren't maintained in stable memory - they are on the Motoko heap.)
 
-Stable memory layout (between upgrades) :
+Stable memory layout (between upgrades), assuming optional stable variable encoding `v_opt`.
 
 ```
-match stable_vars, mem[0..max-1] with
-  None, mem[0..max-1] ->
-    if ic.stable_size() == 0 then
-    assert max == 0;
-    []
-    else
-    // stable grow mem to page including address | 4 + 4 + max - 1 |
-    [word(0x0)@@max@@mem[4 + 4, ..., max - 1]@@mem[0, ..., 4 + 4 -1 ]
-  Some v, mem[0..max-1] ->
-    let blob = serialise(v) in
-    // stable grow page including address | 4 + |blob| + 4 + max |
-    [size(serialize(v))@@blob@@max@@mem[4 + |blob| + 4, ..., max -1 ]@@mem[0, ..., 4 + |blob| + 4  -1]
+(case !size == 0)
+  (case v == None)
+  <empty> (ie. ic0.stable_size() == 0)
+  (case v_opt = Some v)
+  [N..N+3] StableVariable data len
+  [N+4..(N+4)+len-1] StableVariable data
+  [(N+4)+len-1,..M-1] 0...0 // zero padding
+(case !size > 0)
+[0..3]  0...0
+[4..N-1]  StableMemory bytes
+[N..N+3]  StableVariable data len
+[N+4..(N+4)+len-1] StableVariable data
+[(N+4)+len..M-3] 0...0 // zero padding
+[M-3..M-6] value N/64Ki = !size
+[M-5..M-2] saved StableMemory bytes
+[M-1]  version byte
```

Comment:
On second thought, why be greedy instead of aligning to a full word.

Reply:
You mean for the version? Yeah, I agree.

Reply:
I'll leave 4 bytes for the version.

Reply:
Done